compiler/rustc_hir/src/hir.rs

```diff
@@ -942,6 +945,178 @@ pub struct ParentedNode<'tcx> {
     pub node: Node<'tcx>,
 }
 
+/// Arguments passed to an attribute macro.
+#[derive(Clone, Debug, HashStable_Generic, Encodable, Decodable)]
+pub enum AttrArgs {
+    /// No arguments: `#[attr]`.
+    Empty,
+    /// Delimited arguments: `#[attr()/[]/{}]`.
+    Delimited(DelimArgs),
+    /// Arguments of a key-value attribute: `#[attr = "value"]`.
+    Eq {
+        /// Span of the `=` token.
+        eq_span: Span,
+        /// The "value".
+        value: MetaItemLit,
+    },
+}
+
+#[derive(Clone, Debug, Encodable, Decodable)]
+pub enum AttrKind {
```

Comment:
Where do you plan to put the new variants for lowered built-in attributes, into this enum or into something else (`AttrItem`, etc)?

In the first case `AttrItem` can stay `ast::AttrItem`.

In the second case this structure is identical to `ast::AttrKind`, except it uses a different `AttrItem`, so you could parameterize `ast::AttrKind` over its item type and use `type AttrKind = ast::AttrKind<hir::AttrItem>` here instead.
(Similarly to my `Attribute` suggestion below.)

Reply:
This does make sense I think, but the PR would get bigger. I think this should be done in a follow-up?